tensorflow/python/ops/linalg_grad.py

```diff
@@ -613,6 +613,59 @@ def _MatrixTriangularSolveGrad(op, grad):
   return (grad_a, grad_b)
 
 
+# To avoid nan in cases with degenrate eigenvalues or
+# degenrate/zero singular values in calculations of
+# f and s_inv_mat, we introduce a Lorentz brodening.
+def _SafeReciprocal(x, epsilon=1E-20):
+  return x * math_ops.reciprocal(x * x + epsilon)
+
+@ops.RegisterGradient("Eig")
+def _EigGrad(op, grad_e, grad_v):
+  """Gradient for Eig.
+  Based on eq. 4.77 from paper by
+  Christoph Boeddeker et al.
+  https://arxiv.org/abs/1701.00392
+  See also
+  "Computation of eigenvalue and eigenvector derivatives
+  for a general complex-valued eigensystem" by Nico van der Aa.
+  As for now only distinct eigenvalue case is considered.
+  """
+  e = op.outputs[0]
+  compute_v = op.get_attr("compute_v")
+  # a = op.inputs[0], which satisfies
+  # a[...,:,:] * v[...,:,i] = e[...,i] * v[...,i]
+  with ops.control_dependencies([grad_e, grad_v]):
+    if compute_v:
+      v = op.outputs[1]
+      vt = _linalg.adjoint(v)
+      w = linalg_ops.matrix_inverse(vt)
```

Comment:
Comment here that while vt is formally invertible as long as the original matrix has distinct eigenvalues (i.re. is not defective). However, in practice, vt may be ill-conditioned when eigenvalues are close.

Reply:
I believe this is not true. vt is invertible if original matrix is diagonalizable. Also the condition number of eigenvector matrix does not depend on closeness of eigenvalues (in particular, hermitian matrix with all same eigenvalues still have unitary eigenvector matrix). Separate interesting question is whether we can guarantee low condition number by changing eigenvector norm, and if so, should we do it?

Reply:
A matrix is nondefective (diagonalizable) if and only if it has a complete set of eigenvectors, i.e. vt is full rank.

Reply:
My point about cond(vt) is that I suspect that it goes towards infinity as you approach a non-diagonalizable matrix.